flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTaskWriter.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.iceberg.io.FileAppenderFactory;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFileFactory;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Tasks;
+
+public class RowDataTaskWriter extends BaseTaskWriter<RowData> {
+
+  private static final PartitionKey UNPARTITIONED_KEY = new PartitionKey(PartitionSpec.unpartitioned(), null);
+
+  private final Schema schema;
+  private final PartitionKey partitionKey;
+  private final RowDataWrapper rowDataWrapper;
+  private final List<Integer> equalityFieldIds;
+
+  private final Map<PartitionKey, RowDataDeltaWriter> deltaWriterMap = Maps.newHashMap();
+
+  RowDataTaskWriter(PartitionSpec spec,
+                    FileFormat format,
+                    FileAppenderFactory<RowData> appenderFactory,
+                    OutputFileFactory fileFactory,
+                    FileIO io,
+                    long targetFileSize,
+                    Schema schema,
+                    RowType flinkSchema,
+                    List<Integer> equalityFieldIds) {
+    super(spec, format, appenderFactory, fileFactory, io, targetFileSize);
+
+    this.schema = schema;
+    this.equalityFieldIds = equalityFieldIds;
+
+    this.partitionKey = new PartitionKey(spec, schema);
+    this.rowDataWrapper = new RowDataWrapper(flinkSchema, schema.asStruct());
+  }
+
+  @Override
+  public void write(RowData row) throws IOException {
+    RowDataDeltaWriter deltaWriter;
+
+    if (spec().fields().size() <= 0) {
+      // Create and cache the delta writer if absent for unpartitioned table.
+      deltaWriter = deltaWriterMap.get(UNPARTITIONED_KEY);
+      if (deltaWriter == null) {
+        deltaWriter = new RowDataDeltaWriter(null, equalityFieldIds);
+        deltaWriterMap.put(UNPARTITIONED_KEY, deltaWriter);
+      }
+    } else {
+      // Refresh the current partition key.
+      partitionKey.partition(rowDataWrapper.wrap(row));
+
+      // Create and cache the delta writer if absent for partitioned table.
+      deltaWriter = deltaWriterMap.get(partitionKey);
+      if (deltaWriter == null) {
+        // NOTICE: we need to copy a new partition key here, in case of messing up the keys in writers.
+        PartitionKey copiedKey = partitionKey.copy();
+        deltaWriter = new RowDataDeltaWriter(copiedKey, equalityFieldIds);
+        deltaWriterMap.put(copiedKey, deltaWriter);
+      }
+    }
+
+    switch (row.getRowKind()) {
+      case INSERT:
+      case UPDATE_AFTER:
+        deltaWriter.write(row);
+        break;
+
+      case DELETE:
+      case UPDATE_BEFORE:
+        deltaWriter.delete(row);
+        break;
+
+      default:
+        throw new UnsupportedOperationException("Unknown row kind: " + row.getRowKind());
+    }
+  }
+
+  @Override
+  public void close() {
+    Tasks.foreach(deltaWriterMap.values())
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(deltaWriter -> {
```

Comment:
You can supply a checked exception class that will be thrown. That way, you can move the try/catch outside of the `run` block:

```java
  @Override
  public void close() {
    try {
      Tasks.foreach(deltaWriterMap.values())
          .throwFailureWhenFinished()
          .noRetry()
          .run(RowDataDeltaWriter::close);
    } catch (IOException e) {
      throw new UncheckedIOException("Failed to close writers", e);
    }
  }
```

And you can also add a callback that receives the exception for each failure with `onFailure` if you want to log the exceptions individually.